### PR TITLE
NAS-107214 / 12.1 / cache dmidecode results to reduce CPU time

### PR DESCRIPTION
**Pull request by yocalebo**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/system.py [hunk 1]
```diff
@@ -357,10 +357,63 @@ def autotune(self, conf='loader'):
...
+        if self.DMIDECODE_CACHE is None:
+            self.DMIDECODE_CACHE = {
+                'ecc-memory': None,
+                'baseboard-manufacturer': None,
+                'baseboard-product-name': None,
+                'system-manufacturer': None,
+                'system-product-name': None,
+                'system-serial-number': None,
+                'system-version': None,
+            }
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/system.py:382]: Why dont we set that initially in the class definition?

File: src/middlewared/middlewared/plugins/system.py [hunk 2]
```diff
@@ -632,10 +668,13 @@ async def feature_enabled(self, name):
 
     @private
     async def _system_serial(self):
-        return (await(await Popen(
-            ['dmidecode', '-s', 'system-serial-number'],
-            stdout=subprocess.PIPE,
-        )).communicate())[0].decode().strip() or None
+
+        serial = (await self.middleware.call('system.dmidecode_info'))['system-serial']
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/system.py:672]: `return (await self.middleware.call('system.dmidecode_info'))['system-serial'] or None`

is an idea

File: src/freenas/etc/netcli [hunk 3]
```diff
@@ -1477,15 +1477,19 @@ def show_ip():
...
-    proc = subprocess.Popen(['dmidecode', '-s', 'system-serial-number'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding='utf8')
-    serial = proc.communicate()[0].strip()
+
+    product = ''
+    serial = ''
+    version = ''
+
     try:
         with Client() as c:
+            product = c.call('system.dmidecode_info')['system-product-name']
```
Comment (william-gr) [inline src/freenas/etc/netcli:1487]: May want to call system.dmidecode_info only once

File: src/middlewared/middlewared/plugins/system.py [hunk 4]
```diff
@@ -370,8 +369,8 @@ async def dmidecode_info(self):
         dmidecode data is mostly static so cache the results
         """
 
-        if self.DMIDECODE_CACHE is None:
-            self.DMIDECODE_CACHE = {
+        if self.__dmidecode_cache is None:
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/system.py:372]: I meant, why not have DMIDECODE_CACHE = {
                'ecc-memory': None,
                'baseboard-manufacturer': None,
                'baseboard-product-name': None,
                'system-manufacturer': None,
                'system-product-name': None,
                'system-serial-number': None,
                'system-version': None,
            }

in there directly, or now

self.__dmidecode_cache = {
                'ecc-memory': None,
                'baseboard-manufacturer': None,
                'baseboard-product-name': None,
                'system-manufacturer': None,
                'system-product-name': None,
                'system-serial-number': None,
                'system-version': None,
            }

in init? Why do that later and check for None?

Comment (yocalebo) [reply src/middlewared/middlewared/plugins/system.py]: Misunderstood on moving it to init, sorry. I've removed the extra None check.

**Comment by william-gr** [comment]
@yocalebo prefer we dont backport this for 12.0

**Pull request merged**